### PR TITLE
Spreadsheet: Make Range(s).forEach() return Position objects & Add Range(s).at(ix)

### PR DESCRIPTION
**Pull request by u9g**

### CODE REVIEW

File: Base/res/js/Spreadsheet/runtime.js [hunk 1]
```diff
@@ -213,6 +225,19 @@ class Range {
         }
     }
 
+    at(wantedIx) {
+        let ix = 0;
+        let found = null;
+        console.log(this);
```
Comment (Smrtnyk) [inline Base/res/js/Spreadsheet/runtime.js:231]: you forgot to remove this?

File: Base/res/js/Spreadsheet/runtime.js [hunk 2]
```diff
@@ -147,6 +147,18 @@ class Ranges {
         }
     }
 
+    at(wantedIx) {
+        let ix = 0;
+        let found = null;
+        this.forEach(cell => {
```
Comment (Smrtnyk) [inline Base/res/js/Spreadsheet/runtime.js:153]: does libjs support something like `Array.prototype.findIndex` or `Array.prototype.find`?

Comment (u9g) [reply Base/res/js/Spreadsheet/runtime.js:153]: where would you get the array to do it on?

File: Base/res/js/Spreadsheet/runtime.js (hunk 1 above)
Comment (u9g) [reply Base/res/js/Spreadsheet/runtime.js]: good catch, fixed :)

File: Userland/Applications/Spreadsheet/Tests/basic.js [hunk 3]
```diff
@@ -108,4 +108,17 @@ describe("Range", () => {
         expect(R`A0:A25`.first().name).toEqual("A0");
         expect(R`A2:A25`.first().name).toEqual("A2");
     });
+
+    test("Range#at", () => {
```
Comment (alimpfard) [inline Userland/Applications/Spreadsheet/Tests/basic.js:112]: Does JS use this notation? AFAIK it's mostly Ruby?

File: Base/res/js/Spreadsheet/runtime.js [hunk 4]
```diff
@@ -517,35 +511,24 @@ function internal_lookup(
...
-    lookup_inputs.forEach(name => {
-        value = thisSheet[name];
-        if (matches(value)) {
-            didMatch = true;
-            matchingName = name;
-            return Break;
-        }
-        ++i;
-    });
+    i = resolve(lookup_inputs).findIndex(val => matches(val));
```
Comment (alimpfard) [inline Base/res/js/Spreadsheet/runtime.js:515]: This pulls all input values into an array, that's definitely not desired.
e.g. a very large range will require a very large underlying array with this:
with A0: 0, A1...A1000000: 1, you'd end up allocating an array with a million entries even if the needed element was `0`.

**Review by alimpfard — CHANGES_REQUESTED** [review]

File: Base/res/js/Spreadsheet/runtime.js (hunk 4 above)
Comment (u9g) [reply Base/res/js/Spreadsheet/runtime.js:515]: I think I’ll add a find then to the range class

File: Userland/Applications/Spreadsheet/Tests/basic.js (hunk 3 above)
Comment (u9g) [reply Userland/Applications/Spreadsheet/Tests/basic.js:112]: java uses it too, I think its just generic knowledge that static functions are Class.Fn() and instance funcs are Class#Fn()

File: Base/res/js/Spreadsheet/runtime.js (hunk 4 above)
Comment (u9g) [reply Base/res/js/Spreadsheet/runtime.js]: I see what you mean, I brought back most of the code from master, this isn't a valid concern anymore.

File: Base/res/js/Spreadsheet/runtime.js [hunk 5]
```diff
@@ -517,35 +511,24 @@ function internal_lookup(
...
-    if (!didMatch) return if_missing;
+    if (i === -1) return if_missing;
+    matchingName = lookup_inputs.at(i).name;
 
     if (lookup_outputs === undefined) {
         if (reference) return Position.from_name(matchingName);
 
-        return value;
+        return thisSheet[matchingName];
     }
```
[the remark this answers is not in the document]
Comment (u9g) [reply Base/res/js/Spreadsheet/runtime.js]: I see what you mean, I brought back most of the code from master, this isn't a valid concern anymore.

File: Userland/Applications/Spreadsheet/Tests/free-functions.js [hunk 6]
```diff
@@ -44,26 +44,26 @@ describe("Basic functions", () => {
...
-        expect(numericReduce(acc => acc + 1, 0, [])).toEqual(0);
+        expect(numericReduce((acc, x) => acc + x, 19, R`A0:A2`)).toEqual(22);
+        expect(numericReduce(acc => acc + 1, 3, R`A0:A0`)).toEqual(4);
     });
 
     test("numericResolve", () => {
         expect(numericResolve).toBeDefined();
-        expect(numericResolve(["A0", "A1", "A2"])).toEqual([0, 1, 2]);
-        expect(numericResolve([])).toEqual([]);
+        expect(JSON.stringify(numericResolve(R`A0:A2`))).toEqual("[0,1,2]");
```
[the remark this answers is not in the document]
Comment (u9g) [reply Userland/Applications/Spreadsheet/Tests/free-functions.js]: fixed

File: Userland/Applications/Spreadsheet/Tests/free-functions.js [hunk 7]
```diff
@@ -44,26 +44,26 @@ describe("Basic functions", () => {
 
     test("reduce", () => {
         expect(reduce).toBeDefined();
-        expect(reduce(acc => acc + 1, 0, [1, 2, 3, 4])).toEqual(4);
```
[the remark this answers is not in the document]
Comment (u9g) [reply Userland/Applications/Spreadsheet/Tests/free-functions.js:47]: if you want I can do an instanceof for ranges so this still works tho.

[the remark this answers is not in the document]
Comment (alimpfard) [reply Userland/Applications/Spreadsheet/Tests/free-functions.js:47]: Hmmm yeah, let's keep that.

[the remark this answers is not in the document]
Comment (u9g) [reply Userland/Applications/Spreadsheet/Tests/free-functions.js:47]: Okay, I kept the other tests working :)

**Comment by alimpfard** [comment]
Looks like the tests are failing though ^^'

**Pull request merged**